src/blockrelay/mempool_sync.cpp

```diff
@@ -0,0 +1,444 @@
+        LOG(MPOOLSYNC, "Requesting to sync %d missing transactions from %s\n", setHashesToRequest.size(),
+            pfrom->GetLogName());
+
+        return true;
+    }
+
+    // If there are no transactions to request, then synchronization is complete
+    mempoolSyncRequested[pfrom].completed = true;
+
+    LOG(MPOOLSYNC, "Completeing mempool sync with %s; no missing transactions\n", pfrom->GetLogName());
+
+    return true;
+}
+
+bool CRequestMempoolSyncTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
+{
+    CRequestMempoolSyncTx reqMempoolSyncTx;
+    vRecv >> reqMempoolSyncTx;
+
+    // Message consistency checking
+    if (reqMempoolSyncTx.setCheapHashesToRequest.empty())
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Incorrectly constructed getmemsynctx received.  Banning peer=%s", pfrom->GetLogName());
+    }
+
+    // A response was received for a request that was not made
+    if (mempoolSyncResponded.count(pfrom) == 0)
+    {
+        dosMan.Misbehaving(pfrom, 10);
+        return error("Received getmemsynctx from peer %s but mempool sync is not in progress", pfrom->GetLogName());
+    }
+
+    // Already processed requested transactions
+    if (mempoolSyncResponded[pfrom].completed)
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Received getmemsynctx from peer %s but mempool sync has already completed", pfrom->GetLogName());
+    }
+
+    LOG(MPOOLSYNC, "Received getmemsynctx from peer=%s requesting %d transactions\n", pfrom->GetLogName(),
+        reqMempoolSyncTx.setCheapHashesToRequest.size());
+
+    std::vector<uint256> mempoolTxHashes;
+    GetMempoolTxHashes(mempoolTxHashes);
+
+    // Locate transactions requested
+    // Note that only those still in the mempool will be located
+    std::vector<CTransactionRef> vTx;
+    for (auto &hash : mempoolTxHashes)
+    {
+        uint64_t cheapHash = GetShortID(
+            mempoolSyncResponded[pfrom].shorttxidk0, mempoolSyncResponded[pfrom].shorttxidk1, hash, SHORT_ID_VERSION);
+
+        if (reqMempoolSyncTx.setCheapHashesToRequest.count(cheapHash) == 0)
+            continue;
+
+        auto txRef = mempool.get(hash);
+        if (txRef == nullptr)
+            continue;
+
+        vTx.push_back(txRef);
+    }
+
+    LOG(MPOOLSYNC, "Sending %d mempool sync transactions to peer=%s\n", vTx.size(), pfrom->GetLogName());
+
+    // Assemble missing transaction object
+    CMempoolSyncTx mempoolSyncTx(vTx);
+    pfrom->PushMessage(NetMsgType::MEMPOOLSYNCTX, mempoolSyncTx);
+
+    // We should not receive any future messages related to this synchronization round
+    mempoolSyncResponded[pfrom].completed = true;
+
+    return true;
+}
+
+bool CMempoolSyncTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
+{
+    std::string strCommand = NetMsgType::MEMPOOLSYNCTX;
+    CMempoolSyncTx mempoolSyncTx;
+    vRecv >> mempoolSyncTx;
+
+    // Do not process unrequested memsynctx.
+    if (mempoolSyncRequested.count(pfrom) == 0)
+    {
```

Comment:
where are the locks on cs_mempoolsync?  if they are implied before calling this function then add an AssertLockHeld(cs_mempoolsync) at the beginning of the method.

Same for all your other methods, i think there are a few more above here that I don't see any locking for.

Reply:
Good catch. Actually, the `mempoolSyncRequested` and `mempoolSyncResponded` structures were not being locked at all. The only thing that I used `cs_mempoolsync` for was to prevent multiple sync requests from being made simultaneously from ThreadHandler in net.cpp. I went ahead and changed that mutex to `cs_mempoolsync_init` and am now using `cs_mempoolsync` to lock the maps above in mempool_sync.cpp. I figured that it's best to keep the lock activated in net.cpp as lightweight as possible since it's in what I assume to be a potential bottleneck in the code.